### PR TITLE
firefox-gnome-theme: init at 131

### PR DESCRIPTION
**Pull request by ASHGOLDOFFICIAL**
## Package info
**Project description**

A GNOME theme for Firefox

This theme follows latest GNOME Adwaita style.

**Metadata**

* homepage URL: https://github.com/rafaelmardojai/firefox-gnome-theme
* license: unlicense
* platforms: all

**Screenshot**
Screenshot from 2024-05-24 13-31-12

## Basic usage (home-manager)
I don't know how to use it system wide without overlays, wrappers and overrides. 
```nix
programs.firefox.profiles.<name> = {
  settings = {
    # Needed
    # https://github.com/rafaelmardojai/firefox-gnome-theme?tab=readme-ov-file#required-firefox-preferences
    "toolkit.legacyUserProfileCustomizations.stylesheets" = true;
    "svg.context-properties.content.enabled" = true;
    
    # Non essential
    "browser.theme.dark-private-windows" = false;
    "browser.uidensity" = 0;
    "widget.gtk.rounded-bottom-corners.enabled" = true;

    # Theme options can be found at
    # https://github.com/rafaelmardojai/firefox-gnome-theme?tab=readme-ov-file#enabling-optional-features
  };
  userChrome = ''
    @import "${pkgs.firefox-gnome-theme}/userChrome.css";
  '';
  userContent = ''
    @import "${pkgs.firefox-gnome-theme}/userContent.css";
  '';
};
```
For theme to automatically swich between light/dark variants based on GNOME's "Dark Style" option you need to use "System theme — auto" theme in Firefox.

## Questions and what can be done
- Should install scripts be added as well? I think they only affect home directory.
- Maybe add `version` and `hash` arguments for easier override (in case someone need other version of the theme for an older Firefox). Please give me your opinion.
- There are different themes in this package (standard Adwaita and some maia theme), user can choose one in install script. So should there be a `theme` argument for an easier override?
- Maybe there should be home-manager module (`programs.firefox-gnome-theme`) for easier usage. I don't know what can be done via system configuration, so give me your opinion.
- Just a question: `cp` and `install` are used during the `installPhase` so can this package's platforms be set to all? Or should the `installPhase` be moved to `buildPhase`?

## Things done
- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by Aleksanaa — COMMENTED** [review]
Looks okay, but how should people apply this theme as a Nix package?

File: pkgs/by-name/fi/firefox-gnome-theme/package.nix [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+  strictDeps = true;
+
+  # Only copy necessary files
+  installPhase = ''
+    runHook preInstall
+
+    install -d $out
+    cp -ra ./configuration ./theme ./icon.svg ./userChrome.css ./userContent.css -t $out
+    install -Dm 644 ./README.md -t $doc/share/doc/${finalAttr.pname}/
```
Comment (Aleksanaa) [inline pkgs/by-name/fi/firefox-gnome-theme/package.nix:27]:
```suggestion
    install -Dm 644 ./README.md -t $doc/share/doc/firefox-gnome-theme/
```

**Comment by ASHGOLDOFFICIAL** [comment]
If someone can squash these commits, I would be happy. I cannot do it myself because my slow and very restricted university Internet prevent me from cloning and pulling repo without going into efforts, and I don't think it can be done through GitHub website.

**Comment by Aleksanaa** [comment]
I think this package is okay in terms of packaging themes, but I'm not sure whether we should accept this package (and this packaging method), because it seems that this structure cannot be directly used by a configuration tool. So actually we can't decide how to place these files.

**Comment by ASHGOLDOFFICIAL** [comment]
> I think this package is okay in terms of packaging themes, but I'm not sure whether we should accept this package (and this packaging method), because it seems that this structure cannot be directly used by a configuration tool. So actually we can't decide how to place these files.

This theme is basically a set of static CSS files, so is there some preferred way to package and use static files in nixpkgs? How is it normally done?

**Comment by ASHGOLDOFFICIAL** [comment]
Also, there's my work-in-progress home-manager module to simplify its usage. I would like to add `preferences` option to set some theme-specific options in Firefox. Maybe some assertion to check if some needed preferences aren't enabled/disabled in some other place. I also in progress of testing use cases.
```nix
{ config, lib, pkgs, ... }:

let
  cfg = config.custom.firefox-gnome-theme;
  firefox-gnome-theme = pkgs.callPackage ../firefox-gnome-theme.nix {};
in {
  options.custom.firefox-gnome-theme = {
    enable = lib.mkEnableOption "a GNOME theme for Firefox";
    profiles = lib.mkOption {
      default = [];
      description = "List of profile names for which to apply the theme.";
      example = [ "default" ];
      type = with lib.types; listOf str;
    };
    theme = lib.mkOption {
      default = "default";
      description = "Set the colors used in the theme.";
      example = "maia";
      type = lib.types.enum [ "default" "maia" ];
    };
  };

  config = lib.mkIf cfg.enable {
    programs.firefox.profiles = builtins.listToAttrs (map (name: {
      name = name;
      value = let
        importColorchemes = theme: ''
          @import "${firefox-gnome-theme}/theme/colors/light-${theme}.css";
          @import "${firefox-gnome-theme}/theme/colors/dark-${theme}.css";
        '';
      in {
        settings = {
          "toolkit.legacyUserProfileCustomizations.stylesheets" = true;
          "svg.context-properties.content.enabled" = true;
          "browser.theme.dark-private-windows" = lib.mkDefault false;
          "browser.uidensity" = lib.mkDefault 0;
          "widget.gtk.rounded-bottom-corners.enabled" = lib.mkDefault true;
        };

        userChrome = lib.mkBefore (''
          @import "${firefox-gnome-theme}/userChrome.css";
        '' + importColorchemes cfg.theme);

        userContent = lib.mkBefore (''
          @import "${firefox-gnome-theme}/userContent.css";
        '' + importColorchemes cfg.theme);
      };
    }) cfg.profiles);
  };
}

```

**Comment by nyabinary** [comment]
Can this be squashed into one commit?

**Comment by ASHGOLDOFFICIAL** [comment]
Squashed all commits into one, updated the version. But I don't know how to pass `nixfmt-check`. I've run the command, but there are no changes.

**Comment by nyabinary** [comment]
> Squashed all commits into one, updated the version. But I don't know how to pass `nixfmt-check`. I've run the command, but there are no changes.

If you're using flakes
`nix run nixpkgs#nixfmt-rfc-style {path to package.nix here}`

**Comment by ASHGOLDOFFICIAL** [comment]
> Also, for the description of the package, please follow the conventions:
> https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

Would "GNOME theme for Firefox" be a proper description and grammatically correct one?

**Comment by nyabinary** [comment]
> > Also, for the description of the package, please follow the conventions:
> > https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes
> 
> Would "GNOME theme for Firefox" be a proper description and grammatically correct one?

Yeah that should be fine

**Comment by nyabinary** [comment]
DId you forget to format it? :P

**Comment by ASHGOLDOFFICIAL** [comment]
#347611. Recreated PR because I messed up.

**Pull request closed without merging**